2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1]);
+
+    let reader = io::BufReader::new(file.unwrap());
+    let mut current = String::from("");
+    let mut passports = Vec::new();
+    for line in reader.lines() {
+        let l = line.unwrap();
+        if l.len() == 0 {
+            let passport = Passport::from_string(&current);
+            passports.push(passport);
+            current = String::from("");
+        } else {
+            current = current + " " + &l;
+        }
+    }
+    let passport = Passport::from_string(&current);
+    passports.push(passport);
+
+    println!("Found {} total passports", passports.len());
+    println!(
+        "Found {} valid passports",
+        passports.iter().filter(|p| p.is_valid()).count()
+    );
+}
+
+pub struct Passport {
+    birth_year: Option<String>,
+    issue_year: Option<String>,
+    expiration_year: Option<String>,
+    height: Option<String>,
+    hair_color: Option<String>,
+    eye_color: Option<String>,
+    passport_id: Option<String>,
+    country_id: Option<String>,
+}
+
+impl Passport {
+    pub fn from_string(data: &str) -> Passport {
+        let mut birth_year = None;
+        let mut issue_year = None;
+        let mut expiration_year = None;
+        let mut height = None;
+        let mut hair_color = None;
+        let mut eye_color = None;
+        let mut passport_id = None;
+        let mut country_id = None;
+        let parts = data.split_whitespace();
+        for p in parts {
+            let mut items = p.split(':');
+            let name = items.next().unwrap();
+            let value = items.next().unwrap();
+            match name {
+                "byr" => {
+                    birth_year = Some(String::from(value));
```

Comment:
Rust type inference is ✨ magic ✨ and much more powerful from C#'s. Since further down you lock in the type of `birth_year` (and friends) when you assign it to the field in `Passport`, you can take advantage of that elsewhere (you already see that above with the `let mut birth_year = None`).

Rust has a pair of helpful traits (i.e. interfaces) for type conversion, and that combined with the power of the trait system makes for some cool stuff. Types can implement `From<T>` to be convertable from a `T`. Similarly, there's `Into<T>` which converts a type **into** a `T`. Plus, because Rust allows "blanket" implementations, there are automatic implementations that implement `U: Into<T>` for types where `T: From<U>` and vice-versa. 

The `String` type implements `From<&str>`, which means there is a corresponding implementation of `Into<String>` on `&str`. So all that is to say you can actually write this:

All that complexity is to say you can actually just use this:

```suggestion
                    birth_year = Some(value.into());
```

(https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a6b08ee2478840b02332aa0216e7f749)